lazy_dataset/core.py

```diff
@@ -1851,28 +1911,77 @@ def indexable(self):
 
     def __iter__(self):
         buckets = list()
-        for i, sample in enumerate(self.input_dataset):
-            value = min(self.key(sample), self.max_value)
+        for i, example in enumerate(self.input_dataset):
+            value = self.key(example)
             found_bucket = False
-            for i, (bucket, min_value, max_value) in enumerate(buckets):
-                if min_value <= value <= max_value:
-                    bucket.append(sample)
-                    if len(bucket) >= self.batch_size:
-                        buckets.pop(i)
+            for j, (
+                    bucket, creation_idx, lower_bound, upper_bound, max_value
+            ) in enumerate(buckets):
+                if lower_bound <= value <= upper_bound:
+                    bucket.append(example)
+                    max_value = max(max_value, value)
+                    if (
+                            len(bucket) >= self.batch_size
+                            or (
+                            self.total_size_threshold is not None
+                            and (len(
+                        bucket) * max_value) > self.total_size_threshold
```

Comment:
When `total_size_threshold` is used, the batch if yielded, when `len(bucket)*max_length_in_bucket > total_size_threshold`. Can you document, that the batch size is bigger than the threshold?
In extreme cases it is approximately `2 * max_length_in_bucket`.